### PR TITLE
chore(RHTAPWATCH-973): functional tests

### PR DESCRIPTION
**Pull request by Omeramsc**
Adds functional tests for the different endpoints in the workspace-manager webserver

### CODE REVIEW

File: cmd/main_test.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+	resp, err := performHTTPGetCall(url, header)
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Unexpected error testing the \"%s\" endpoint: %v", url, err))
+	Expect(resp.StatusCode).To(Equal(expectedCode))
+	Expect(strings.TrimSpace(expectedBody)).To(Equal(strings.TrimSpace(resp.Body)))
+},
+	Entry(
+		"Calling the workspace endpoint for user1 responds only with the 'test-tenant' workspace info",
+		HTTPheader{"X-Email", "user1@konflux.dev"},
+		http.StatusOK,
+		`{"kind":"WorkspaceList","apiVersion":"toolchain.dev.openshift.com/v1alpha1","metadata":{},"items":[{"kind":"Workspace","apiVersion":"toolchain.dev.openshift.com/v1alpha1","metadata":{"name":"test-tenant","creationTimestamp":null},"status":{"namespaces":[{"name":"test-tenant","type":"default"}]}}]}`),
```
Comment (yftacherzog) [inline cmd/main_test.go]: Is it possible to add line breaks to those payloads to make them more readable?

File: cmd/main_test.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	Entry(
+		"Workspace endpoint for user2 responds with 2 namespaces info",
+		HTTPheader{"X-Email", "user2@konflux.dev"},
+		http.StatusOK,
+		`{"kind":"WorkspaceList","apiVersion":"toolchain.dev.openshift.com/v1alpha1","metadata":{},"items":[{"kind":"Workspace","apiVersion":"toolchain.dev.openshift.com/v1alpha1","metadata":{"name":"test-tenant","creationTimestamp":null},"status":{"namespaces":[{"name":"test-tenant","type":"default"}]}},{"kind":"Workspace","apiVersion":"toolchain.dev.openshift.com/v1alpha1","metadata":{"name":"test-tenant-2","creationTimestamp":null},"status":{"namespaces":[{"name":"test-tenant-2","type":"default"}]}}]}`),
+	Entry(
+		"Workspace endpoint with no header",
+		HTTPheader{},
+		500,
+		`{"message":"Internal Server Error"}`),
```
Comment (yftacherzog) [inline cmd/main_test.go:180]: I think we should open an issue for this (here on the repo). I think return code 500 usually means a bug.
Can you do that?
You can keep the test case.

File: cmd/main_test.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+)
```
Comment (yftacherzog) [inline cmd/main_test.go:181]: I think we should add another test case for header containing a user that should not see any workspace.

**Review by yftacherzog — COMMENTED** [review]
Good job. Left some comments, but otherwise lgtm.

File: cmd/main_test.go (hunk 1 above)
Comment (Omeramsc) [reply cmd/main_test.go]: fixed

File: cmd/main_test.go (hunk 2 above)
Comment (Omeramsc) [reply cmd/main_test.go:180]: will open

File: cmd/main_test.go (hunk 3 above)
Comment (Omeramsc) [reply cmd/main_test.go:181]: fixed

File: cmd/main_test.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+			`"metadata":{"name":"test-tenant","creationTimestamp":null},"status":`+
+			`{"namespaces":[{"name":"test-tenant","type":"default"}]}}]}`),
+	Entry(
+		"Workspace endpoint for user2 responds with 2 namespaces info",
+		HTTPheader{"X-Email", "user2@konflux.dev"},
+		http.StatusOK,
+		`{"kind":"WorkspaceList","apiVersion":"toolchain.dev.openshift.com/v1alpha1","metadata":{},`+
+			`"items":[{"kind":"Workspace","apiVersion":"toolchain.dev.openshift.com/v1alpha1",`+
+			`"metadata":{"name":"test-tenant","creationTimestamp":null},"status":{"namespaces":`+
+			`[{"name":"test-tenant","type":"default"}]}},{"kind":"Workspace","apiVersion":"toolchain.dev.openshift.com/v1alpha1"`+
```
Comment (yftacherzog) [inline cmd/main_test.go]: Can we make this line nice as well?

Comment (Omeramsc) [reply cmd/main_test.go]: I find it not as nice when broken to more lines, but fixed.

**Comment by yftacherzog** [comment]
looks good! One last thing, did you check that the tests you added fail the CI when one of them fails?

File: cmd/main_test.go (hunk 2 above)
Comment (Omeramsc) [reply cmd/main_test.go:180]: https://issues.redhat.com/browse/RHTAPWATCH-1026

**Comment by Omeramsc** [comment]
> looks good! One last thing, did you check that the tests you added fail the CI when one of them fails?

Yes, when an entry fails, the test results are as follows:
```
Ran 7 of 7 Specs in 7.350 seconds
FAIL! -- 6 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestCmd (7.35s)
FAIL
coverage: 1.1% of statements
FAIL    github.com/konflux-ci/workspace-manager/cmd     7.379s
FAIL
make: *** [Makefile:65: test] Error 1
```

File: cmd/main_test.go (hunk 2 above)
Comment (yftacherzog) [reply cmd/main_test.go:180]: I meant creating an issue on the GitHub repo.

**Comment by yftacherzog** [comment]
> > looks good! One last thing, did you check that the tests you added fail the CI when one of them fails?
> 
> Yes, when an entry fails, the test results are as follows:
> 
> ```
> Ran 7 of 7 Specs in 7.350 seconds
> FAIL! -- 6 Passed | 1 Failed | 0 Pending | 0 Skipped
> --- FAIL: TestCmd (7.35s)
> FAIL
> coverage: 1.1% of statements
> FAIL    github.com/konflux-ci/workspace-manager/cmd     7.379s
> FAIL
> make: *** [Makefile:65: test] Error 1
> ```

Is this in CLI or in CI on the repo?

**Comment by Omeramsc** [comment]
> > > looks good! One last thing, did you check that the tests you added fail the CI when one of them fails?
> > 
> > 
> > Yes, when an entry fails, the test results are as follows:
> > ```
> > Ran 7 of 7 Specs in 7.350 seconds
> > FAIL! -- 6 Passed | 1 Failed | 0 Pending | 0 Skipped
> > --- FAIL: TestCmd (7.35s)
> > FAIL
> > coverage: 1.1% of statements
> > FAIL    github.com/konflux-ci/workspace-manager/cmd     7.379s
> > FAIL
> > make: *** [Makefile:65: test] Error 1
> > ```
> 
> Is this in CLI or in CI on the repo?

CLI, those tests are yet to fail in the PR (besides the first iteration of it, which did fail the CI). 
are we expecting any unusual responses from `go test` or `ginkgo` ?

**Comment by yftacherzog** [comment]
> > > > looks good! One last thing, did you check that the tests you added fail the CI when one of them fails?
> > > 
> > > 
> > > Yes, when an entry fails, the test results are as follows:
> > > ```
> > > Ran 7 of 7 Specs in 7.350 seconds
> > > FAIL! -- 6 Passed | 1 Failed | 0 Pending | 0 Skipped
> > > --- FAIL: TestCmd (7.35s)
> > > FAIL
> > > coverage: 1.1% of statements
> > > FAIL    github.com/konflux-ci/workspace-manager/cmd     7.379s
> > > FAIL
> > > make: *** [Makefile:65: test] Error 1
> > > ```
> > 
> > 
> > Is this in CLI or in CI on the repo?
> 
> CLI, those tests are yet to fail in the PR (besides the first iteration of it, which did fail the CI). are we expecting any unusual responses from `go test` or `ginkgo` ?

I just want to see that we're adding tests that can fail the CI in case of regressions. Can you try and push a broken test (e.g. change the value of an assertion) and confirm the CI is failing?

**Comment by yftacherzog** [comment]
I see it's failing as expected. Let's fix it, and we can merge.

**Comment by Omeramsc** [comment]
@yftacherzog Sure, confirmed
image

**Review by yftacherzog — APPROVED** [review]
good job. lgtm

**Pull request merged**